Display remaining time

JoeCohen opened the pull request:
Displays (estimated) time remaining for iNat import.
- Prepares for importing all a user's iNat obss.  
(IMO this feature (or similar) is a necessity for a good User Experience when importing large numbers of observations.)
- Resolves #3003 

### Manual Test
- Create several disposable iNat observations
- Import them, paying attention to the Remaining Time 
Expected result:
Remaining time should
  - Appear
  - Change as time elapses
  - Finish as "00:00:00" when the import is Done
- Delete the disposable iNat observations

JoeCohen committed:
Draft display of remaining time

NOTE: gets `Errors: NoMethodError: undefined method `paginate' for nil` when importing multiple Obss

JoeCohen committed:
Create new provisionals after Name.parse

- Fixes a bug where an error was thrown if the API tried to create a Name when a close match already existed.
- Uses Name.parse on iNat provisionals
- Creates a new MO Name iff there's MO provisional with that text_name
- Uses parsed iNat provisional to create the new Name

JoeCohen committed:
Update expectation

JoeCohen committed:
Fix failure from prior commit

JoeCohen committed:
Improve job_tracker_test

- Test elapsed time
- `elapsed` => `elapsed_time` to parallel `remaining_time`
- padded hours for elapsed time display
- DRY job_tracker

JoeCohen committed:
Merge branch 'main' into jdc-3003-display-remaining-time

mo-nathan commented on app/jobs/inat_import_job.rb:
```diff
@@ -231,19 +231,20 @@ def new_obs_params
   def id_or_provisional_or_species_name
-    prov_name = @inat_obs.provisional_name
-    return @inat_obs.name_id if prov_name.blank?
+    return @inat_obs.name_id if @inat_obs.provisional_name.blank?
```
Not a big deal, but why is this change in this PR?

mo-nathan commented:
The "remaining time" just went up and showed the same values as the elapsed time until it was done.  I just did two observations, but it seemed pretty surprising.

mo-nathan commented on app/models/inat_import_job_tracker.rb:
```diff
@@ -22,17 +22,25 @@ def status
+    remaining_importables = importables - imported_count
+    cumulative_avg_import_time = elapsed_seconds / imported_count
+    seconds =
+      (remaining_importables * cumulative_avg_import_time).to_i
```
I don't see anything in this calculation that takes into account the time spent on the current import.  I would expect something like:
```
total_expected_time = total_count * avg_import_time
remaining_time = total_expected_time - elapsed_seconds
```
where `avg_import_time` only gets updated after each import completes.

mo-nathan reviewed (changes requested):
I saw surprising behavior when I imported just two observations where the remaining time tracked the total time until the process was done.  I think I see the issue and added a comment.

JoeCohen commented:
Thanks for the review and comments.

JoeCohen committed:
Merge branch 'main' into jdc-3003-display-remaining-time

JoeCohen committed:
Add inat_import #total_time and #total_import columns

JoeCohen committed:
Initially estimate importables if user lists them

JoeCohen committed:
Update InatImport comments

- Add comments r ne attributes
- Clarify other comments

JoeCohen committed:
Increment total_imports count

JoeCohen committed:
"Remaining Time" => "Estimated Remaining Time"

JoeCohen committed:
Separate time calculations from their presentations

JoeCohen committed:
Update import total_time when job is done

JoeCohen committed:
Rename `total_imports` => `total_imported_count`

For consistency with `imported_count`

JoeCohen committed:
Update InatImport Attributes comments

JoeCohen committed:
Update total_time => total seconds in code

Forgot to do this whe I renamed the column

JoeCohen committed:
explanatory comment

JoeCohen committed:
Calculate total_expected_time

JoeCohen committed:
Fix InatImport tests and coverage

JoeCohen committed:
Add InatImports#avg_import_time column

JoeCohen committed:
Set initial avg_import_time

JoeCohen committed:
Update remaining time between imports

- Responds to [@mo-nathan's request](https://github.com/MushroomObserver/mushroom-observer/pull/3004#discussion_r2132937632)
- uses `roy_inat_import` fixture instead of funky `timings_import`
- changes `elapsed_seconds` to `elapsed_time`

JoeCohen committed:
Add attribute & methods to time one import

- Adds InatImport#last_obs_start attribute
- Adds methods to get elapsed time of last obs, reset last_obs_start

JoeCohen committed:
Revise documentation

JoeCohen committed:
Improve JobTest's manual creation of InatImport

- Improve documentation
- Include all attributes

JoeCohen committed:
DRY create_inat_import

- Pass any param to `create_inat_import` via `create_ivars_from_filename`

JoeCohen committed:
Update import timings after each iNat obs

JoeCohen committed:
Fix Estimated Remaining Time

- Fixes #3046

JoeCohen committed:
Make time_in_hours_minutes_seconds into a helper

JoeCohen committed:
improve comment

JoeCohen commented on app/models/inat_import.rb:
```diff
@@ -41,7 +54,52 @@ def add_response_error(error)
+    end
+  end
+
+  def reset_last_obs_start
+    update(last_obs_start: Time.now.utc)
+  end
+
+  def last_obs_elapsed_time
+    return 0 unless last_obs_start
+
+    Time.now.utc - last_obs_start
+  end
```
@mo-nathan: Was thinking about handling this via a small `StopWatch` class with `elapsed_time` and `reset` methods.
Reactions?

mo-nathan replied:
Sounds worth trying, but not urgent.

JoeCohen committed:
Add error message explaining assertion

JoeCohen committed:
Remove model method which was moved to helper

JoeCohen committed:
Translation string instead of raw English string

Delivers #3051

JoeCohen commented:
I will later this evening try to add missing coverage to InatImportsController. I don't want to hold up the PR any more.

JoeCohen commented:
@mo-nathan, @nimmolo 
Any suggestions for covering `/app/controllers/inat_imports/job_trackers_controller.rb#show`, which is just a Turbo endpoint

JoeCohen committed:
Cover missing lines of InatImportsController

JoeCohen commented:
> Update: `assert_turbo_stream` does not seem to work as i described.
> 
> It checks for a `<turbo-stream>` custom HTML element, and I don't think that shows up in the response unless explicitly rendered.

Thanks for checking on that.
I vaguely recall Nathan doing something like a render for Field Slips in order to test it. I'll look at that later.

mo-nathan commented:
Search for `format: :turbo_stream` in the tests for examples of what I did in order to test turbo stuff in controllers.

JoeCohen committed:
Cover InatImportJobTracker

JoeCohen commented:
Thanks @nimmolo & @nathan. 
I complete covered the turbo endpoint via `get(:show, ... format: :turbo_stream)`, then checking `@response.body` for the expected displayed text.

mo-nathan reviewed (approved):
Looks good!  Thanks for all the work on this.
